src/lib/SaferNames/Syntax.hs

```diff
@@ -0,0 +1,585 @@
+                          deriving Show
+
+type Var = AnnVarP Type :: E
+type Binder     = AnnBinderP PlainBinder            Type  :: B
+type BinderList = AnnBinderP PlainBinderList (ListE Type) :: B
+
+data DataDef n where
+  DataDef :: RawName -> BinderList n l -> ListE DataConDef l -> DataDef n
+
+data DataConDef n = DataConDef RawName (EmptyNest Binder n)
+                    deriving Show
+
+data NamedDataDef n = NamedDataDef (Name n) (DataDef n)
+                      deriving Show
+
+data Block n where Block :: Nest Decl n l -> Expr l -> Block n
+
+type LamExpr = Abs Binder (WithArrow Block)  :: E
+type PiType  = Abs Binder (WithArrow Type)   :: E
+
+data WithArrow (e::E) (n::S) = WithArrow (Arrow n) (e n)
+                     deriving Show
+
+withoutArrow :: WithArrow e n -> e n
+withoutArrow (WithArrow _ x) = x
+
+justArrow :: WithArrow e n -> Arrow n
+justArrow (WithArrow arr _) = arr
+
+type Arrow n = ArrowP (EffectRow n)
+
+type Val  = Atom
+type Type = Atom
+type Kind = Type
+
+type TC  n = PrimTC  (Atom n)
+type Con n = PrimCon (Atom n)
+type Op  n = PrimOp  (Atom n)
+type Hof n = PrimHof (Atom n)
+
+-- === top-level modules ===
+
+type SourceName = T.Text
+
+data SourceNameMap n = SourceNameMap
+  { fromSourceNameMap :: M.Map SourceName (Name n)}
+
+data Module n where
+  Module :: IRVariant
+         -> Nest Decl n l
+         -> BindingsFrag l l'
+         -> SourceNameMap l'
+         -> Module n
+
+type WithBindings = Abs BindingsFrag
+type EvaluatedModule = WithBindings SourceNameMap  :: E
+
+emptyEvaluatedModule :: EvaluatedModule n
+emptyEvaluatedModule = Abs (RecEnvFrag emptyEnv) mempty
+
+type IndexStructure = Nest Binder
+
+getBaseMonoidType :: Scope n -> Type n -> Type n
+getBaseMonoidType scope ty = case ty of
+  TabTy i b -> case projectNamesL scope (i@>()) b of
+    Just b' -> b'
+    Nothing -> error "Can't use a dependent table as a monoid"
+  _         -> ty
+
+-- === effects ===
+
+data EffectRow n = EffectRow (S.Set (Effect n)) (Maybe (Name n))
+                   deriving Show
+
+data Effect n = RWSEffect RWS (Name n) | ExceptionEffect | IOEffect
+                deriving (Show, Eq, Ord)
+
+pattern Pure :: EffectRow n
+pattern Pure <- ((\(EffectRow effs t) -> (S.null effs, t)) -> (True, Nothing))
+ where  Pure = EffectRow mempty Nothing
+
+-- === Recursive bindings ===
+
+data BinderInfo n =
+   LamBound (ArrowP ())
+ -- TODO: make the expression optional, for when it's effectful?
+ -- (or we could put the effect tag on the let annotation)
+ | PatBound
+ | LetBound LetAnn (Expr n)
+ | PiBound
+ | UnknownBinder
+ | DataDefName (DataDef n)
+ | ClassName -- TODO
+ | MethodName  (NamedDataDef n) Int  -- TODO: just have a class name and a label
+ | DataConName (NamedDataDef n) Int
+ | TypeConName (NamedDataDef n)
+ -- TODO: avoid this case by having the inference monad use an alternative
+ -- version of BinderInfo
+ | InferenceName
+
+data TypedBinderInfo n = TypedBinderInfo (Maybe (Type n)) (BinderInfo n)
```

Comment:
It's because things there are bindings for things that aren't atom-like, like the `DataDefName`, that don't have names. There are other solutions, like carrying the "static info payload" type on the name itself. But that's for the next rewrite, I think.

Reply:
(I'll add a note)